[코일] API 테스트/문서자동화 미션 제출합니다.

slowCoyle opened the pull request:
잘부탁드립니다! :)

Ramen6315 committed:
docs : README.md 파일 작성

Commit:
docs: 1단계 README.md 작성

Ramen6315 committed:
feat: 회원 가입 화면 연결

Commit:
setting: Basic, SessionInterceptor 제거

Commit:
feat: 회원 가입 API 연결

Ramen6315 committed:
setting: Gson Gradle 추가

Ramen6315 committed:
test: 회원 가입 성공 테스트 구현

Commit:
test: 회원 가입 유효성 검증 테스트

Ramen6315 committed:
refactor: MemberRequest field 어노테이션 수정

Ramen6315 committed:
feat: api 문서 자동화 구현

Commit:
test: JwtTokenProviderTest 작성

Ramen6315 committed:
feat: 로그인 기능 구현

Commit:
feat: Login 이메일 비밀번호 유효성 확인

Commit:
feat: Login 이메일 비밀번호 포맷 유효성 확인

Commit:
feat: Login 이메일 비밀번호 포맷 유효성 확인

Commit:
feat: 더미 데이터 삽입 ApplicationRunner

Ramen6315 committed:
test: 로그인 테스트 코드 작성

Commit:
setting: test application.properties 생성

Commit:
feat: Login Rest docs 생성

Ramen6315 committed:
feat: session 저장소에 token 저장

Ramen6315 committed:
test: Service Test 작성

Commit:
docs: 회원 정보 조회,수정,삭제 기능 목록 작성

Commit:
feat: 회원 정보 조회, 수정, 삭제 화면 연결

Ramen6315 committed:
test: 정보 조회 테스트 구현

Ramen6315 committed:
feat: BearerAuthInterceptor preHandle 구현

Ramen6315 committed:
refactor: InvalidAuthenticationException 생성자 수정

Ramen6315 committed:
refactor: 불필요 주석 삭제

Commit:
fix: 회원가입시 토큰 검증 불필요 로직 분기

BearerAuthInterceptor /Members 로 접근하는 모든 요청을
Interceptor에서 토큰 유효성 검증을 하여 발생하는 에러를
POST일 경우 건너뛰게 변경

Ramen6315 committed:
test: 회원 정보 수정 테스트 구현

Commit:
refactor: 세션에 email을 저장하도록 변경

 - Login.js 세션에 email을 저장하도록 변경
 - LoginMemberController 로그인 반환 DTO 변경

Commit:
feat: 마이 페이지 정보 조회 API 연결

Commit:
refactor: updateMember 메서드 반환값 변경

Commit:
feat: UpdateMember argument resolver 추가

Commit:
feat: member 이메일 동등성 확인 메서드 추가

Commit:
feat: update API 권한 확인 로직 추가

Commit:
fix: 회원 정보 가져오기

 - LoginMemberMethodArgumentResolver session에 저장된
 키값의 이름 변경

Commit:
feat: 내 정보 수정 API 연결

Commit:
feat: 회원 삭제 기능 구현 및 테스트 코드 작성

Commit:
fix: 회원 수정 로직 변경

mock 메서드 추가

Commit:
fix: Controller Advice 적용 시 발생 테스트 수정

Commit:
feat: 회원 삭제 기능 API 연결

Commit:
feat: README.md 2단계 즐겨찾기 기능목록 작성

Commit:
feat: FAVORITE table 추가

Commit:
setting: front sample file 추가

Commit:
feat: 즐겨 찾기 인수 테스트 작성

Commit:
feat: 즐겨 찾기 추가 테스트 작성

Commit:
feat: 즐겨 찾기 추가 기능 구현

Commit:
feat: Favorite 추가 인증 기능 구현

Ramen6315 committed:
feat: 즐겨 찾기 추가 Controller Test작성 및 구현

Ramen6315 committed:
test: 회원이 아닌 이메일 JWT 토큰 헤더 요청 시 즐겨찾기 추가 요청 테스트 변경

Ramen6315 committed:
feat: 즐겨찾기 추가 API 연결

Ramen6315 committed:
test: 즐겨 찾기 조회 테스트 코드 작성

Ramen6315 committed:
feat: Controller, Service, 예외 처리 테스트 코드 작성

Ramen6315 committed:
feat: snackbar.js 추가

Ramen6315 committed:
setting: Favorite 더미데이터 Configuration 추가

Ramen6315 committed:
fix: member 조회 NPE 해결

Ramen6315 committed:
refactor: FavoriteResponse ID 추가

개별적인 Favorite 요청을 식별하기 위해 ID 값을 DTO에 추가

Ramen6315 committed:
feat: Favorite 조회 API 연결 및 뷰 생성

Ramen6315 committed:
refactor: FavoriteDeleteAPI가 Id를 받도록 테스트 코드 수정

 - 데이터를 식별 할 수 있도록 Id로 Delete 변경

Ramen6315 committed:
refactor: FavoriteDeleteAPI 가 Id를 요청 받도록 Production code 변경

Ramen6315 committed:
refactor: FavoriteDeleteAPI 뷰 연결

Ramen6315 committed:
feat: README 파일 갱신

Ramen6315 committed:
feat: 회원탈퇴 api docs 추가

Ramen6315 committed:
feat: 회원 수정 api docs 추가

Ramen6315 committed:
feat: 회원 조회 api docs 추가

Ramen6315 committed:
feat: 즐겨 찾기 조회 api docs 추가

Ramen6315 committed:
feat: 즐겨 찾기 추가 api docs 추가

Ramen6315 committed:
feat: 즐겨 찾기 삭제 api docs 추가

Ramen6315 committed:
refactor: 코드 수정

Ramen6315 committed:
fix: 멤버 인수 테스트 수정

 - 헤더를 받지 않던 요청에 헤더 추가

jaeyeonling commented on README.md:
```diff
@@ -0,0 +1,143 @@
+     - [x] 인증 토큰이 없을 경우
+     - [x] 권한이 없는 토큰일 경우
+         
+    - [ ] API 연결
```
해당 부분 체크를 놓치셨네요!

slowCoyle replied:
수정하였습니다! :)

jaeyeonling commented on build.gradle:
```diff
@@ -13,6 +14,9 @@ repositories {
 dependencies {
+	asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor:2.0.4.RELEASE'
+	testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc:2.0.4.RELEASE'
+	implementation 'com.google.code.gson:gson:2.8.6'
```
gson과 jackson은 동일한 기능을 제공하고 있습니다.
jackson이 있는 상황에서 gson을 추가로 사용하신 이유가 있을까요?

slowCoyle replied:
jackson의 존재를 모르고 사용해왔던 gson을 자연스럽게 사용했네요! :) jackson을 알게되었으니 모두 수정하도록 하겠습니다! 감사합니다! 👍

jaeyeonling commented on build.gradle:
```diff
@@ -13,6 +14,9 @@ repositories {
 }
 
 dependencies {
+	asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor:2.0.4.RELEASE'
+	testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc:2.0.4.RELEASE'
```
```suggestion
	asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor'
	testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
```

plugins에서 'io.spring.dependency-management'가 어떠한 역할을 하는지 공부해보세요~

slowCoyle replied:
'io.spring.dependency-management'가 사용하고 있는 Spring boot 버전에 맞춰 검증된 버전으로 의존성을 가져오네요! 감사합니다!

jaeyeonling commented on src/main/java/wooteco/subway/config/WebMvcConfig.java:
```diff
@@ -3,39 +3,44 @@
-        registry.addInterceptor(sessionInterceptor).addPathPatterns("/me/session");
-        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/me/bearer");
+        registry.addInterceptor(bearerAuthMemberInterceptor).addPathPatterns("/members", "/members/*");
+        registry.addInterceptor(bearerAuthFavoriteInterceptor).addPathPatterns("/favorites", "/favorites/*", "/favorites*");
```
![image](https://user-images.githubusercontent.com/40811999/83041202-304d1100-a07b-11ea-9fbd-d54768156e9e.png)

패턴 양식을 확인해주세여~

slowCoyle replied:
auth가 필요한 uri는 모두 한 곳으로 처리하기 위해 `/auth/**`로 처리하였습니다!

jaeyeonling commented on src/main/java/wooteco/subway/domain/favorite/Favorite.java:
```diff
@@ -0,0 +1,57 @@
+    @Id
+    private Long id;
+    private String memberEmail;
+    private String source;
+    private String target;
```
이름을 가지고 있는다면 이름이 변경됐을 경우 올바르지 않게 동작할 수 있습니다.
식별자를 가지도록 수정해보세요.

slowCoyle replied:
source와 target, memberEmail도 의존하고 있는 엔티티의 id 값으로 수정하였습니다.

역 이름 혹은 멤버의 이메일로 접근하면 DB 접근량을 줄일 수 있을 것이라 생각하였는데, 기본키로 해당하는 데이터로 찾는 편이 오히려 유일성, 속도 면에서 더 뛰어날 수 있겠네요!

jaeyeonling commented on src/main/java/wooteco/subway/domain/favorite/Favorite.java:
```diff
@@ -0,0 +1,57 @@
+public class Favorite {
+    @Id
+    private Long id;
+    private String memberEmail;
```
해당 정보 또한 email이 아닌 식별자로 변경해보세요!

slowCoyle replied:
네! 위와 마찬가지로 변경하였습니다! 조언 감사합니다! :)

jaeyeonling commented on src/main/java/wooteco/subway/service/favorite/FavoriteService.java:
```diff
@@ -0,0 +1,50 @@
+        this.favoriteRepository = favoriteRepository;
+    }
+
+    public Favorite createFavorite(String source, String target, String memberEmail) {
+        return favoriteRepository.save(new Favorite(source, target, memberEmail));
+    }
+
+    public Favorite findFavoriteBySourceAndTarget(String source, String target) {
+        return favoriteRepository.findBySourceAndTarget(source, target)
+                .orElseThrow(() -> new NotExistFavoriteDataException(source + "," + target));
+    }
```
해당 메서드는 테스트 코드에서만 사용하는데 의도한게 맞을까요?

slowCoyle replied:
역 id로 값을 찾는만큼 필요하지 않아 제거하였습니다! :)

jaeyeonling commented on src/main/java/wooteco/subway/web/member/argumentresolver/FavoriteMethodArgumentResolver.java:
```diff
@@ -0,0 +1,53 @@
+import static org.springframework.web.context.request.RequestAttributes.SCOPE_REQUEST;
+
+@Component
+public class FavoriteMethodArgumentResolver implements HandlerMethodArgumentResolver {
```
해당 기능은 ArgumentResolver의 역할이 아닌 Service의 역할로 보여집니다.
요청이 왔을 때 인증에 대한 부분은 MethodArgumentResolver로 처리하고, 비즈니스 로직이나 로직 순서에 관련된 역할은 Service로 이동시켜주세요!

slowCoyle replied:
이게 저희 페어가 구현하면서 잘못생각했었던 것 같습니다. ArgumentResolver에 너무 많은 역할을 두려하니 추상화가 어려워 여러 ArgumentResolver들이 생겨났고 모두 관리해주기 어려워졌던 것 같습니다. 오히려 Service에서 해야하는 토큰 인증 외의 것들을 ArgumentResolver에서 하려하니 코드가 난잡해졌습니다. 지금은 하나의 ArgumentResolver, Interceptor로 인가를 관리해줍니다! 

추후 ArgumentResolver를 사용해야한다면, DB에서 알맞은 Object로 매핑해주는 포매터의 역할로만 사용하겠습니다! 
감사합니다! :)

jaeyeonling commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -21,31 +23,36 @@ public Member createMember(Member member) {
 
-    public void updateMember(Long id, UpdateMemberRequest param) {
+    public Long updateMember(Long id, UpdateMemberRequest param) {
         Member member = memberRepository.findById(id).orElseThrow(RuntimeException::new);
```
RuntimeException 만으로는 정확한 의도 전달을 할 수 있을까요?
클라이언트에게 정확한 예외처리 메시지를 전달해주는 것이 좋지 않을까요!?

jaeyeonling commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -21,31 +23,36 @@ public Member createMember(Member member) {
-        return member.checkPassword(password);
+    public Member findMemberById(Long id) {
+        return memberRepository.findById(id)
+                .orElseThrow(() -> new NotExistMemberDataException(String.valueOf(id)));
```
NotExistMemberDataException은 email을 매개변수로 받는데, id를 넣고 있네요.
의도한 내용이 맞을까요?

slowCoyle replied:
수정하였습니다! 매개 변수로 예외가 발생한 이유가 무엇인지를 나타낼 수 있는 메시지로 변경했습니다!
id가 문제라면 "id = ", email이 문제라면 "email = "와 같이 변경하였습니다.

jaeyeonling commented on src/main/java/wooteco/subway/service/member/dto/MemberRequest.java:
```diff
@@ -2,11 +2,27 @@
     private String email;
+    @NotBlank(message = "이름에 빈값을 넣지 마세요.")
     private String name;
+    @NotBlank(message = "빈값을 넣지 마세요.")
```
LoginRequest와 동일한 검증을 해야하지 않을까요?

slowCoyle replied:
여러 validation annotaion을 사용해보려하다 신이 나버렸네요...!

리팩토링을 할 때 확인했어야하는데...!

일관성있게 처리할 수 있도록 노력하겠습니다! :)

jaeyeonling commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,46 @@
+
+    @PostMapping("/favorites")
+    public ResponseEntity<FavoriteResponse> createFavorite(@LoginMember Member member, @RequestBody CreateFavoriteRequest request) {
+        Favorite present = favoriteService.save(request, member.getEmail());
```
present는 의도를 나타내는 올바른 변수명이 맞을까요?

slowCoyle replied:
현재 저장돼있는 값을 다룬다고 생각해서 present라는 변수명으로 작성하였는데요!

더 명확한 변수명을 위해 createdFavorites로 변경하였습니다!

jaeyeonling commented on src/main/java/wooteco/subway/web/member/exception/InvalidPasswordException.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
+public class InvalidPasswordException extends IllegalArgumentException{
```
```suggestion
public class InvalidPasswordException extends IllegalArgumentException {
```

IntelliJ Auto Formatting 기능을 사용하는 습관을 들여보아요.

The auto-formatting shortcut in IntelliJ

* For windows Ctrl + Alt + L.
* For ubuntu Ctrl + Alt + windows + L.
* For Mac Command + Option + L.

slowCoyle replied:
윽 죄송합니다...! 신경썼어야하는데...! 다음에는 더 신경써서 이런 일 없도록 하겠습니다 ㅠ 거듭 죄송합니다 ㅠㅜ

jaeyeonling commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthFavoriteInterceptor.java:
```diff
@@ -0,0 +1,68 @@
+
+
+        if (isDelete(request)) {
+            final Map<String, String> pathVariables = (Map<String, String>) request
```
@SuppressWarnings("unchecked")

가능한 경고는 제거해주세요~

slowCoyle replied:
감사합니다! :)

jaeyeonling commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthFavoriteInterceptor.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Map;
+
+@Component
+public class BearerAuthFavoriteInterceptor implements HandlerInterceptor {
```
Auth는 Member에 대한 인증일 것으로 예상이 되는데요.
AuthInterceptor는 Favorite와 Member가 나눠져 있는 이유가 있을까요?

slowCoyle replied:
ArgumentResolver가 가지고 있는 역할이 크다보니, 추상화가 잘 되지 않고 Favorite과 Member로 접근하는 path마다 다른 처리가 필요하다고 판단되어서 나눴습니다.

그러나, ArgumentResolver의 책임을 줄여 추상화를 하니, Interceptor를 하나로 통일 시켜 Member의 인증을 관리할 수 있었습니다.

jaeyeonling commented on src/main/java/wooteco/subway/web/member/interceptor/BearerAuthMemberInterceptor.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest request,
```
회원가입과 로그인에 대한 path를 분리한다면 해당 로직이 조금 간소화될 것 같네요.

slowCoyle replied:
인증이 필요한 path와 인가가 필요한 path를 분리하였습니다! :)

jaeyeonling commented on src/test/java/wooteco/subway/acceptance/favorite/FavoriteAcceptanceTest.java:
```diff
@@ -0,0 +1,114 @@
+
+    private void deleteFavorite(FavoriteResponse favoriteResponse, String token) {
+        given().
+                header("Authorization", "Bearer " + token).
```
```suggestion
                header(AuthorizationExtractor.AUTHORIZATION, "Bearer " + token).
```

가능하면 의미있는 상수를 사용하는 습관을 들여주세요~

slowCoyle replied:
오! 좋은 팁 감사합니다 매번 상수화를 하는 것이 좋을까 고민도 했었거든요!
모두 수정하였습니다!

jaeyeonling commented on README.md:
```diff
@@ -0,0 +1,143 @@
+# 즐겨찾기 - API 테스트/문서자동화
```
예상하지 못한 예외에 대한 처리도 있으면 좋을 것 같습니다.
유저에게는 서버에서 예상한 응답만 전달하도록 해주세요!

방법은 `ControllerAdvice`와 `ExceptionHandler` 키워드로 검색해보세요!

slowCoyle replied:
예상하지 못한 예외까지 처리할 수 있도록 ControllerAdvice를 작성하였습니다!:)

jaeyeonling commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @DeleteMapping("/favorites/{id}")
+    public ResponseEntity<Void> deleteFavorite(@CreateFavorite Favorite favorite) {
```
PathVariable은 어디에 있을까요!?

slowCoyle replied:
ArgumentResolver에 너무 많은 역할을 두다 보니 PathVariable을 사용하여 해당하는 데이터를 조회해오는 책임도 ArgumentResolver가 맡게 되면서 제거되었네요...!

리팩토링하면서 수정하였습니다~!

jaeyeonling commented on src/main/java/wooteco/subway/web/FavoriteController.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @GetMapping("/favorites")
+    public ResponseEntity<FavoritesResponse> createFavorite(@LoginMember Member member) {
```
해당 메서드의 기능과 메서드명이 일치하지 않습니다!

slowCoyle replied:
죄송합니다ㅠ 주의하도록 하겠습니다...!

jaeyeonling commented on src/main/java/wooteco/subway/service/favorite/FavoriteService.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.stream.Collectors;
+
+@Service
+public class FavoriteService {
```
JDBC 식으로 변경하기 위해서는 해당 위치에서 
Member 조회 후 Member.addFavorite(), Member.removeFavorite() 등 메서드를 만들어 내부 컬렉션에 저장하도록 수정하면 되겠네요!

jaeyeonling commented on src/test/java/wooteco/subway/web/favorite/FavoriteControllerIntegrationTest.java:
```diff
@@ -0,0 +1,236 @@
+@ExtendWith(RestDocumentationExtension.class)
+@SpringBootTest
+@AutoConfigureMockMvc
+public class FavoriteControllerIntegrationTest {
```
해당 클래스가 굉장히 많은 의존성을 가지고 있네요!
Mock 또한 굉장히 많이 사용되고 있는데요.
가능한 Mock을 제거하고 의존성을 줄여보면 어떨까요?

jaeyeonling commented on src/test/java/wooteco/subway/web/favorite/FavoriteControllerIntegrationTest.java:
```diff
@@ -0,0 +1,236 @@
+                .header("Authorization", "Bearer" + unExistEmailToken)
+                .accept(MediaType.APPLICATION_JSON))
+                .andDo(print())
+                .andExpect(status().isUnauthorized());
```
```suggestion
                .andExpect(status().isForbidden());
```

Unauthorized(401)와 Forbidden(403)의 차이를 공부해보세요!

slowCoyle replied:
401은 익명의 사용자(로그인 X), 403는 로그인은 하였으나 권한이 없는 사용자가 되겠군요!

jaeyeonling reviewed (changes requested):
안녕하세요 코일!
요구사항에 맞게 잘 구현해주셨습니다 👍 
피드백 남겨드렸으니 확인 후 반영해주세요!

Commit:
refactor: 인증,인가 로직 추상화

 - Interceptor -
인증, 인가를 불필요하게 여러 Interceptor로 처리하던 것을 하나의 Interceptor로 단일화시킴

 - ArgumentResolver -
처음에 의도 했던 ArgumentResolver보다 많은 책임을 가지고 있었고 그 떄문에 파생된 ArgumentResolver들이 생성됨, 그래서 ArgumentResolver가 가지고 있던 요청 멤버가 그 요청을 처리할 수 있는지 확인하는 절차를 제거히여 의도 했던 ArgumentResolver를 생성하고 단일화 시킴.

 - Path Url -
인가가 필요한 요청은 /auth의 경로로 묶어 관리하였다.

 - Sql -
즐겨찾기를 의미하는 FAVORITE 테이블의 관계를 파악하기 쉽도록 모든 관계 컬럼들을 해당 테이블의 id로 설

Commit:
refactor: 인증,인가 테스트 수정

인증,인가 로직 추상화에서 발생한 테스트 실패를 성공시켰다.

Commit:
refactor: gson 의존성 제거

gson을 모두 jackson으로 변경하였다.

Commit:
refactor: 사용하지 않는 interceptor, argumentresolver 삭제

Commit:
refactor: Exception 처리 메시지 변경

및 그래들 버전 제거

Commit:
refactor: LoginRequest, MemberRequest 동일한 유효성 검증 전략으로 수정

Commit:
refactor: 공통 Exception 처리

Advice Controller로 공통된 Exception을 처리하고 ErrorResponse를 통해 공통 규격으로
에러를 반환하도록 변경하였다.

Commit:
refactor: Java Convention 적용

사용하지 않는 메서드 제거

Commit:
refactor: 인증 관련 예외 AuthenticationException 상속 변경

Commit:
refactor: 인증 관련 예외 AuthenticationException 상속 변경에 따른 메소드 처리

Commit:
refactor: Member와 Favorite의 종속 관계화

Member가 Favorite의 List를 가지고 있도록 변경하였다. RDB와 OOP의 패러다임 불일치 제거를 위해
 선택적 ORM인 Spring data JDBC를 적극 활용하였다.

또한, Favorites를 List로 두어 즐겨찾기의 순서를 보장하였다.

Commit:
refactor: FavoriteRepository 삭제

사용하지 않은 Repository 삭제

Commit:
refactor: todo 제거

Commit:
test: ArgumentResolver, Interceptor, extreactor 방어적 테스트 작성

Commit:
feat: Favorite Station 외래키 제약 조건 적용

테스트에서 사용하는 trancate.sql 하기 전 외래키 제약 조건을 제거해주는 명령어 추가

Commit:
feat: Station 삭제 시 연관된 Favorite 삭제

DB 마이그레이션을 대비하여 서버단에서도 외래키 제약 조건처럼 이행하도록 변경

Commit:
refactor: 필요없는 출력문 제거

jaeyeonling commented on src/main/java/wooteco/subway/service/station/StationService.java:
```diff
@@ -27,6 +32,18 @@ public Station createStation(Station station) {
 
     public void deleteStationById(Long id) {
```
중간 연산이 실패하였을 때 원자성을 보장하기 위해 `Transaction`을 이용할 수 있습니다.

slowCoyle replied:
오늘 트랜잭션 격리 수준에 대해 공부했는데... 트랜잭션 어노테이션을 깜빡했네요....!

감사드립니다! :)

jaeyeonling commented on src/main/java/wooteco/subway/service/station/StationService.java:
```diff
@@ -27,6 +32,18 @@ public Station createStation(Station station) {
     }
+
+    private void deleteFavoriteIfRelatedToStationId(Long id) {
+        Optional<Favorite> favoriteBySourceId = favoriteRepository.findBySourceId(id);
```
조회 후 삭제할 필요 없이 DELETE 쿼리를 바로 날려도 될 것 같네요!

slowCoyle replied:
조회 후 삭제한 이유는 역과 관련있는 데이터가 없을 경우 DB에 삭제 쿼리를 날리는 비용을 줄여보려했습니다만 DELETE 쿼리 자체가 탐색 후 없으면 아무런 영향이 없군요...!

감사합니다! :)

jaeyeonling commented on src/test/java/wooteco/subway/domain/favorite/FavoriteRepositoryTest.java:
```diff
@@ -0,0 +1,111 @@
+
+@SpringBootTest
+@Sql("/truncate.sql")
+class FavoriteRepositoryTest {
```
Q1.
JDBC Repository 테스트 문제점
Member - Favorite - Station이 관계를 맺고, Station과 Favorite를 외래키로 설정하여 FavoriteRepositoryTest를 작성할 때 MemberRepository, FavoriteRepository, StationRepository를 모두 주입해줘야하는 상황이 발생하는데요!
외래키 제약조건을 건다면 하나의 RepositoryTest에 관련된 Repository를 주입하여 테스트 하는게 맞는 걸까요 ?
이런 상황이 지속된다면 설계가 잘못되었다는 신호가 되는걸까요 ?

A1.
가능한 의존성을 제거해주는 것이 좋지만, 필요하다면 추가해줄 수 밖에 없겠네요 T.T
해당 부분은 현재 상황에서는 어쩔 수 없는 부분으로 보여집니다!
불필요한 의존성을 제거하기 위해 미리 SQL을 통해 테스트 데이터를 추가해주시는 분들도 있습니다 ㅎㅎ


Q2.
외래키 제약조건 마이그레이션 문제점
Station이 삭제될 때 관련된 즐겨찾기도 삭제해야되는 기능 요구사항을 만족시키려 할 때 가장 먼저 생각난 것은 Station과 Favorite를 외래키로 설정하고 onDelete Cascade를 적용시키자 였습니다.
JPA는 위 기능 요구사항을 만족시키기 위해 더 좋은 방법이 있을 것 같지만 Spring Data JDBC에는 관련된 기능이 없는 것 같더라구요!
그래서 외래키로 설정하고 cascade를 설정하니 만약 DB 마이그레이션이 될 경우에는 기능 요구 사항을 만족하지 못하지 않을까 ? 하는 생각이 들어 StationService에서 FavoriteRepository를 주입받고 Station이 제거될 때 연관된 Favorite들을 지울 수 있도록 코드를 작성하였습니다.
사실 위의 과정은 우리가 작성한 Web Server Application에서의 구현만으로 마이그레이션되는 상황까지도 만족시키는 구현이 될 수 있을 것 같더라구요!
근데 다르게 생각해보면 DB단에서도 참조 무결성을 지키기 위해 외래키 제약 조건을 두고 제 어플리케이션에서도 동일한 작업을 해주는 것이 이중 안전 장치로 더 안정적이지 않을까 생각하여 둘 다 유지 시켰습니다.
재연링의 생각이 궁금해요! 하나의 기능을 구현하기 위해 오버 프로그래밍한 것이 아닌가하는 생각이 듭니다...! :smiley:

A2.
저는 DB 설정과 별개로 코드로 작성해주는 것이 좋다고 생각하는데요.
일반적으로 라이브 환경에서 제약조건을 잘 사용하지 않는 곳이 많은데요. 
이유로는 CS 대응, 장애 발생 시 대응 등 라이브 환경에서 제약조건이 장점보다는 단점이 크게 다가오는 것 같습니다.
테스트 환경에서는 제약조건을 건 상태로 테스트 후 라이브에서는 끄는 회사도 봤고
아예 모든 제약조건을 사용 안하는 회사도 봤습니다 ㅎㅎ
예전에 프로시저나, DB 스케줄러 등으로 처리하던 작업들도 서버에서 처리해주는 추세로 흘러가는데요.
이러한 흐름으로 봤을 때 DB에서 처리해주는 작업은 줄어들 것 같아요.
사람마다 다르겠지만 제 결론은...!
`DB와는 별개의 레이어로 생각하고 서버 입장에서 할 수 있는 모든 것을 해주는게 좋다!` 입니다.

slowCoyle replied:
1. 미리 SQL을 통해 테스트 데이터를 추가하는 것도 좋은 대안이 되겠네요! :) 감사합니다!

2. SQL에 제약조건을 걸어두니 테스트 코드의 양도 증가하고 테스트와 직접적으로 관련되지 않은 불필요한 코드들이 증가하더라구요! 답변감사합니다! :)

혹시 라이브 환경이라고 하면 실제 배포되는 코드를 말씀하시는 걸까요 ?

jaeyeonling replied:
네 맞습니다!
실제 배포하여 운영중인 환경을 말합니다!

jaeyeonling reviewed (changes requested):
추가적인 미션을 잘 구현해주셨네요 💯 
질문에 대한 답변 남겨드렸으니 확인해주세요~

Commit:
refactor: Service Layer 트랜잭션화

Commit:
refactor: 불필요한 연산 제거

검색 후 삭제의 의미가 없다고 판단하여 삭제를 바로 실행하도록 변경함

jaeyeonling reviewed (approved):
수 많은 피드백을 반영하시느라 고생 많으셨습니다 💯 
해당 요청은 머지하도록 하겠습니다!
